File: SelfClaw.Infrastructure/Extensions/Plugins/PluginPermissions.cs
namespace SelfClaw.Infrastructure.Extensions.Plugins;

/// <summary>
/// Permissions are a disclosure list the user acknowledges before a Plugin is enabled, so an unknown
/// bare token stays legal — nothing enforces it, and rejecting it would break packages that already
/// declare vocabulary such as <c>workspace.read</c>. A prefixed permission is different:
/// <c>network.fetch:</c> widens the panel CSP directly, so its value is parsed rather than trusted.
/// </summary>
internal static class PluginPermissions
{
    public const string Panel = "ui.panel";
    public const string ContextRead = "host.context.read";
    public const string TranscriptRead = "host.transcript.read";
    public const string ComposerWrite = "host.composer.write";
    public const string WorkspaceRead = "host.workspace.read";
    public const string NetworkFetchPrefix = "network.fetch:";

    private const int MaximumLength = 256;

    /// <summary>
    /// Returns the normalized, distinct, ordinal-sorted permission set. Both the manifest reader and the
    /// enable/acknowledge path run their input through here, so the two can never disagree about whether
    /// a package's permissions changed — a disagreement would leave a Plugin permanently unenablable.
    /// </summary>
    public static IReadOnlyList<string> Validate(IReadOnlyList<string?>? permissions)
    {
        var values = (permissions ?? [])
            .Select(permission => NormalizeOne(permission?.Trim() ?? string.Empty))
            .ToArray();
        if (values.Distinct(StringComparer.Ordinal).Count() != values.Length)
        {
            throw new InvalidDataException("Plugin permissions must be unique.");
        }

        return values.OrderBy(permission => permission, StringComparer.Ordinal).ToArray();
    }

    public static bool Grants(IReadOnlyList<string> permissions, string permission)
        => permissions.Contains(permission, StringComparer.Ordinal);

    /// <summary>
    /// The origins a panel may reach with fetch/XHR/WebSocket. Anything absent is blocked by the panel's
    /// <c>connect-src</c>, so a Plugin that declares nothing is fully offline.
    /// </summary>
    public static IReadOnlyList<string> ReadNetworkOrigins(IReadOnlyList<string> permissions)
        => permissions
            .Where(permission => permission.StartsWith(NetworkFetchPrefix, StringComparison.Ordinal))
            .Select(permission => permission[NetworkFetchPrefix.Length..])
            .Distinct(StringComparer.Ordinal)
            .OrderBy(origin => origin, StringComparer.Ordinal)
            .ToArray();

    private static string NormalizeOne(string permission)
    {
        if (string.IsNullOrWhiteSpace(permission) || permission.Length > MaximumLength)
        {
            throw new InvalidDataException("Plugin permissions contain an invalid value.");
        }

        if (!permission.Contains(':', StringComparison.Ordinal))
        {
            if (permission.Any(character =>
                    !char.IsAsciiLetterOrDigit(character) && character is not '.' and not '-' and not '_'))
            {
                throw new InvalidDataException("Plugin permissions contain an invalid value.");
            }

            return permission;
        }

        if (!permission.StartsWith(NetworkFetchPrefix, StringComparison.Ordinal))
        {
            throw new InvalidDataException($"Plugin permission '{permission}' uses an unsupported prefix.");
        }

        return NetworkFetchPrefix + NormalizeNetworkOrigin(permission[NetworkFetchPrefix.Length..]);
    }

    // The value must reduce to a bare origin, because that is exactly what a CSP source expression
    // accepts. A path, query or credentials would either be dropped silently by the browser or widen the
    // grant past what the acknowledgement dialog showed. Normalizing here (rather than demanding a
    // canonical spelling from the author) keeps "https://Api.Example.com:443/" and
    // "https://api.example.com" from acknowledging as two different permissions.
    private static string NormalizeNetworkOrigin(string origin)
    {
        if (!Uri.TryCreate(origin, UriKind.Absolute, out var uri) ||
            uri.Scheme is not ("https" or "http") ||
            !string.IsNullOrEmpty(uri.UserInfo) ||
            uri.AbsolutePath.Trim('/').Length > 0 ||
            !string.IsNullOrEmpty(uri.Query) ||
            !string.IsNullOrEmpty(uri.Fragment))
        {
            throw new InvalidDataException(
                $"Plugin permission '{NetworkFetchPrefix}{origin}' must be a bare origin such as https://api.example.com.");
        }

        if (uri.Scheme == Uri.UriSchemeHttp && !uri.IsLoopback)
        {
            throw new InvalidDataException(
                $"Plugin permission '{NetworkFetchPrefix}{origin}' must use HTTPS for non-loopback hosts.");
        }

        return uri.GetLeftPart(UriPartial.Authority);
    }
}

File: SelfClaw.TranscriptVue/src/renderers/pluginPermissions.js
// Permission tokens are a disclosure list, so the dialog explains the ones the host
// actually enforces and echoes anything else verbatim rather than hiding it.
const permissionDescriptions = {
	'ui.panel': '在右侧插件面板区域添加自定义界面',
	'host.context.read': '读取当前会话、工作区与选中 Agent 的上下文',
	'host.transcript.read': '读取当前对话的转录内容',
	'host.composer.write': '向消息输入框写入文本',
	'host.workspace.read': '读取当前工作区的信息',
	'hooks.run': '在回合开始/结束时运行插件命令；可读取你的提示词与模型的最终回复，可阻止回合并注入上下文',
	'hooks.tool': '拦截工具调用；可读取工具参数与结果，可拒绝、要求确认、修改参数（完全访问模式下修改后的参数不经确认直接执行）并向模型追加反馈',
	'hooks.http': '观察模型 HTTP 请求与响应的元数据（已脱敏），可追加 x-* 与链路追踪请求头',
	'hooks.http.body': '读取发送给模型提供商的请求体（含对话内容，≤ 1 MiB）',
};

const hookPermissionPrefix = 'hooks.';
const networkFetchPrefix = 'network.fetch:';

export function describePluginPermission(token) {
	if (!token) {
		return '';
	}

	if (token.startsWith(networkFetchPrefix)) {
		return `允许面板向 ${token.slice(networkFetchPrefix.length)} 发起网络请求`;
	}

	return permissionDescriptions[token] || String(token);
}

// hooks.http.body is the only token that hands the plugin the conversation itself,
// so the dialog renders it with the caution treatment instead of a plain row.
export function isSensitivePluginPermission(token) {
	return token === 'hooks.http.body';
}

export function hasHookPermissions(permissions) {
	return (permissions || []).some((permission) => String(permission).startsWith(hookPermissionPrefix));
}
